src/Middleware/HeaderPropagation/samples/HeaderPropagationSample/HeaderPropagationSample/Startup.cs

```diff
@@ -0,0 +1,102 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.HeaderPropagation;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Extensions;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+
+namespace HeaderPropagationSample
+{
+    public class Startup
+    {
+        public void ConfigureServices(IServiceCollection services)
+        {
+            services.AddRouting();
+
+            // A configuration similar to https://github.com/apache/incubator-zipkin-b3-propagation
+            // but without sampling.
+            //
+            // To see this in action, send a request to / that includes headers like:
+            // X-TraceId: 43a3cb27-f494-42c4-a05b-09fb256b9021
+            // X-SpanId: 32756e86-0bef-409b-8dbc-86e6f8ddcea9
+            //
+            // This demonstrates three common uses of header propagation:
+            // 1. Forward a header as-is
+            // 2. Forward a header using a different header name
+            // 3. Generate a new header value, or conditionally generate a header value
+            services.AddHeaderPropagation(options =>
+            {
+                // Propagate the X-TraceId as a representation of the overall transaction.
+                options.Headers.Add("X-TraceId");
+
+                // Propagate the X-SpanId as X-ParentSpanId to represent the 'parent' of this HTTP call.
+                options.Headers.Add(new HeaderPropagationEntry("X-SpanId", "X-ParentSpanId"));
```

Comment:
I considered also adding *sugar* methods for these, but I think we'd want to get creative. The meaning of `Add(string)` is obvious, it's less so with this.  I'd be happy to consider suggestions 😀

Reply:
If these are your three main use cases then I think they do warrant sugar. 
`options.Headers.Add(new HeaderPropagationEntry("X-SpanId", "X-ParentSpanId"));`
is no more specific about what the parameters mean than
`options.Headers.Add("X-SpanId", "X-ParentSpanId");`

It's not like you have a wide variety of overloads that would confuse the meaning of the second string parameter.

Reply:
I fear using this for distributed tracing works against the right way to do it (what we’re baking in). Are there any other reasons to use this besides tracing?

Reply:
> I fear using this for distributed tracing works against the right way to do it (what we’re baking in). Are there any other reasons to use this besides tracing?

I feel like we talked about this before, but I'm sure it's going to come up again every time we look at this code 😆. 

We're building-in one particular implementation of tracing - this is a general building block for *your stuff*. Real things you can use this for:
- Doing a different tracing thing than we're doing.
- Various routing/versioning strategies: https://istio.io/docs/tasks/traffic-management/request-routing/
- A/B testing with service -> service calls
- Fowarding things like a language for resources/messages

I'm happy to show another sample here if you feel like this makes life too confusing. I agree that our guidance should be to use the built-in tracing features once they are available.

Reply:
Yes please add any other sample,one of those sounds good. The problem with having a distributed tracing scenario is that it only 1/2 works. You need to also store things in logs and have other parts of the system use it (outside of ASP.NET
Core) which is why the Activity based imply works well.

Reply:
Hi Ryan,
glad we are on the same page on the possible use cases for this.
I guess adding all those scenarios in the example will help to not get this comment lost. let me know if I can help.


------
David - you are on point mentioning logs, as it is relevant for all the scenarios, not only for tracing. As I started to replace the current logic on applications with this package (as I published a backport package to 2.1-2.2), it has been the main pain point (on top of the clumsy setup which now Ryan got much better :) ).

I was originally tempted to include the ability to also add the header to the logging scope, but it seemed out of scope. However right now to add a value to the logging scope you either have to capture it from the `HttpContext` again on a separate middleware or, expecially when you rely on the factory, you have to access the `HeaderPropagationValues` (which hold these values) from a separate middleware to add those values to the logging scope, which feels wrong (I guess ideally it should be internal but then you will close this door).

Sorry for going off topic, but It would be really nice to have your opinion on this. Thank you.

Reply:
@davidfowl - do you have a doc on what's going on with Activity and logging that we can link here for @alefranz